spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaConsumerBackoffManager.java

```diff
@@ -47,87 +53,162 @@
  */
 public class KafkaConsumerBackoffManager implements ApplicationListener<ListenerContainerPartitionIdleEvent> {
 
+	private static final LogAccessor logger = new LogAccessor(LogFactory.getLog(RetryTopicConfigurer.class));
 	/**
 	 * Internal Back Off Clock Bean Name.
 	 */
 	public static final String INTERNAL_BACKOFF_CLOCK_BEAN_NAME = "internalBackOffClock";
 
+	private static final int TIMING_CORRECTION_THRESHOLD = 100;
+
+	private static final int POLL_TIMEOUTS_FOR_CORRECTION_WINDOW = 2;
+
 	private final ListenerContainerRegistry registry;
 
-	private final Map<TopicPartition, Context> backOffTimes;
+	private final Map<TopicPartition, Context> backOffContexts;
 
 	private final Clock clock;
 
+	private final TaskExecutor taskExecutor;
+
+	private final Sleeper sleeper;
+
 	public KafkaConsumerBackoffManager(ListenerContainerRegistry registry,
-									@Qualifier(INTERNAL_BACKOFF_CLOCK_BEAN_NAME) Clock clock) {
+									@Qualifier(INTERNAL_BACKOFF_CLOCK_BEAN_NAME) Clock clock,
```

Comment:
Since we develop here a library, not the target project, it is better to stay away from auto-wiring in this type of code.
Why do we need `Clock` at all? Why can't we just use a `System.nanoTime()`?

Reply:
This is a trickier one, and one I had my doubts too. The main purpose of injecting the clock is for testing purposes, so we can inject a fixed time clock - easier to hit than a moving target. Same reason to use a Sleeper instead of just Thread sleeping.

I had doubts about the qualifier - maybe we could just inject whatever clock that app uses, if any - and create an instance if none as . The only requirement of course is that it has to be the same clock throughout the backoff process so that it keeps in sync.

That being said, do you think it's better to just use System.millis (or nanos)? I can look into that if so.

Reply:
I think the same kind of reasoning applies to the injected TaskExecutor, which I was looking for a better solution already. What I would do is create a separate TaskExecutor bean in the RetryTopicBootstrapper, qualify the injection here, and then shut it down on a ContextClosedEvent.

I could also create a new instance in the KafkaConsumerBackoffManager, but then it would spawn an actual thread in the unit tests - usually not what we'd want if it was an app.

Reply:
Given the proximity of RC1, tomorrow, I think we can live with this for now; I can scheduled an RC2 if need be.